org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServiceAccessor.java

```diff
@@ -485,6 +484,13 @@ private static LanguageServerWrapper getLSWrapper(@Nullable IProject project,
 		}
 	}
 
+	public static LanguageServerWrapper getLSWrapperForConnection(@NonNull IDocument document,
```

Comment:
@mickaelistria , the main question is we want to expose this new API or something else, I see that the existing `getLSWrapperForConnection` in this package has been marked as deprecated.

Reply:
I used to consider this method as a bad smell, something we'd like to get rid of; so making it API doesn't make me comfortable. However, now that we've established that LanguageServerWrapper is a proper API element, maybe it makes sense now.
But couldn't we instead add something like `LanguageServers.forDocument().withId(languageServerId)` to be more consistent with other operations and get rid of this odd method?

Reply:
So far @ahmedneilhussain and I discussed several times how have an executor with a single server, and it is a bit weird for the given API that expects consumers to deal with n servers, so far we refrained from such executor/filter. I could create such a filter but if it is acceptable to export the method I make now public, I think that is preferable.

Reply:
Yeah, I also wondered about a `forDocument().withId(languageServerId)` style method. But I think it was for this very use-case! When I went through the existing API usage by usage (see document attached https://github.com/eclipse/lsp4e/pull/251#issuecomment-1337648823 ) this class `CommandExecutor` was an awkward outlier.

I think if there's a rare one-off need then a one-off solution is OK. `CommandExecutor` is part of the framework so I think it's OK for it to have privileged access to low-level API that we don't want 'ordinary' users to be doing. It would be nice if this could be package-private but we'd need to move classes around a bit for that to work.

Reply:
Actually you have now made the change I was about to suggest: all that this code is trying to do is make sure that the server that issued a command is the one that executed it.... but in fact we can just do that by storing the wrapper in the code mining. 👏